lib/kube/proxy/forwarder.go

```diff
@@ -2148,3 +2312,358 @@ func (f *Forwarder) removeKubeDetails(name string) {
+			allowedKubeGroups, allowedKubeUsers, err := authCtx.Checker.CheckKubeGroupsAndUsers(
+				authCtx.sessionTTL,
+				false,
+				services.NewKubernetesClusterLabelMatcher(authCtx.kubeClusterLabels),
+				services.NewKubernetesResourceMatcher(
+					types.KubernetesResource{
+						Kind:      types.KindKubePod,
+						Name:      item.Name,
+						Namespace: item.Namespace,
+					},
+				),
+			)
+			// no match was found, we ignore the request.
+			if err != nil {
+				continue
+			}
+			allowedKubeUsers, allowedKubeGroups = fillDefaultKubePrincipalDetails(allowedKubeUsers, allowedKubeGroups, authCtx.User.GetName())
+
+			impersonatedUsers, impersonatedGroups, err := computeImpersonatedPrincipals(
+				f.log, utils.StringsSet(allowedKubeUsers), utils.StringsSet(allowedKubeGroups),
+				req.Header,
+			)
+			if err != nil {
+				continue
+			}
+
+			// create a new kubernetes.Client using the impersonated users and groups
+			// that matched the current pod.
+			client, err := newImpersonatedKubeClient(details.kubeCreds, impersonatedUsers, impersonatedGroups)
+			if err != nil {
+				return internalErrStatus, trace.Wrap(err)
+			}
+			// delete each pod individually.
+			err = client.CoreV1().Pods(item.Namespace).Delete(req.Context(), item.Name, deleteOptions)
+			if err != nil {
+				// TODO(tigrato): check what should we do when delete returns an error.
+				// Should we check if it's permission error?
+				continue
+			}
+			items = append(items, item)
+		}
+		// reset items.
+		o.Items = items
+	default:
+		return internalErrStatus, trace.BadParameter("expected *corev1.PodList, got: %T", obj)
+	}
+	// reset the memory buffer.
+	memWriter.Buffer().Reset()
+	// encode the filtered response into the memory buffer.
+	if err := encoder.Encode(obj, memWriter.Buffer()); err != nil {
+		return internalErrStatus, trace.Wrap(err)
+	}
+	// copy the output into the user's ResponseWriter and return.
+	return memWriter.Status(), trace.Wrap(memWriter.CopyInto(w))
+}
+
+// newImpersonatedKubeClient creates a new Kubernetes Client that impersonates 3
+// a username and the groups.
+func newImpersonatedKubeClient(creds kubeCreds, username string, groups []string) (*kubernetes.Clientset, error) {
+	c := &rest.Config{}
+	// clone cluster's rest config.
+	*c = *creds.getKubeRestConfig()
+	// change the impersonated headers.
+	c.Impersonate = rest.ImpersonationConfig{
+		UserName: username,
+		Groups:   groups,
+	}
+	// TODO(tigrato): reuse the http client.
+	client, err := kubernetes.NewForConfig(c)
+	return client, trace.Wrap(err)
+}
+
+// parseDeleteCollectionBody parses the request body targeted to pod collection
+// endpoints.
+func parseDeleteCollectionBody(r io.Reader, decoder runtime.Decoder) (metav1.DeleteOptions, error) {
+	into := metav1.DeleteOptions{}
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return into, trace.Wrap(err)
+	}
+	if len(data) > 0 {
+		_, _, err = decoder.Decode(data, nil, &into)
+	}
+	return into, trace.Wrap(err)
```

Comment:
Suggestion: easier to see that `err` is handled.

```suggestion
	if len(data) == 0 {
		return into, nil
	}
	_, _, err = decoder.Decode(data, nil, &into)
	return into, trace.Wrap(err)
```